_posts/2022-05-09-project-loom-for-distributed-systems.md

```diff
@@ -0,0 +1,397 @@
+material, and should the results violate the invariants, one has cast iron evidence of a bug. As a comparative tool,
+the results speak for themselves. In the early days, many fanciful claims made by database companies bit the dust, and more recently contracting Kyle Kingsbury to stress your database has become something of a rite of passage.
+
+![jepsen](/assets/images/jepsen.jpg){: width="100%" }
+
+As a white box tool for bug detection, Jepsen is fantastic. If you've written the database in question, Jepsen leaves something to be desired. By falling down to the lowest common denominator of 'the database must run on Linux', testing is both slow and non-deterministic because most production-level actions one can take are comparatively slow. For a quick example, suppose I'm looking for bugs in Apache Cassandra which occur due to adding and removing nodes. It's usual for adding and removing nodes to Cassandra to take hours or even days, although for small databases it might be possible in minutes, probably not much less than. A Jepsen environment could only run one iteration of the test every few minutes; if the failure case only happens one time in every few thousand attempts, without massive parallelism I might expect to discover issues only every few days, if that. I had an improvement that I was testing out against a Cassandra cluster which I discovered deviated from Cassandra's pre-existing behaviour with (against a production workload) probability one in a billion. Jepsen likely won't help with that.
+
+Suppose that we either have a large server farm or a large amount of time and have detected the bug somewhere in our stack of at least tens of thousands of lines of code. We now have a bug, but no reproduction of said bug. If there is some kind of smoking gun in the bug report or a sufficiently small set of potential causes, this might just be the start of an odyssey.
+
+Jepsen is probably the best known example of this type of testing, and it certainly moved the state of the art; most database authors have similar suites of tests. ScyllaDB documents their testing strategy [here](https://www.scylladb.com/product/technology/scylla-testing/) and while the styles of testing might vary between different vendors, the strategis have mostly coalesced around this approach.
+
+Jepsen is limited here by its broadness. As the author of the database, we have much more access to the database if we so desire, as shown by [FoundationDB](https://www.foundationdb.org/).
+
+## FoundationDB
+
+One of the most personally influential tech talks I've watched is
+[Will Wilson's FoundationDB talk at StrangeLoop 2014](https://www.youtube.com/watch?v=4fFDFbi3toc).
```

Comment:
another good article on this subject: https://downtimeproject.com/podcast/talkin-testing-with-sujay-jayakar/

The idea of generating deterministic execution traces for async systems is a bit older than FDB. I remember a Google paper, maybe on Chubby (?), that talked about this, too.